### PR TITLE
Get Rhome and libR from Preferences.jl when provided

### PR DESCRIPTION
**Pull request by frankier**
See: https://github.com/JuliaInterop/RCall.jl/issues/490

### CODE REVIEW

**Comment by palday** [comment]
@frankier I haven't forgotten you! But I also probably won't review this today, my apologies. :frowning: 

Thanks for your contribution.

**Comment by ParadaCarleton** [comment]
This would definitely be super useful! I'd be very happy to see it merged.

File: Project.toml
Comment (palday) [inline Project.toml:14]: needs compat entry

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:32]: I think this blurb should stay in.

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:55]: This might be a good spot to call out another advantage of project-specify R via preferences -- some users might have multiple R versions installed and they can now prefer different R versions for different Julia projects.

File: docs/src/installation.md
Comment (frankier) [reply docs/src/installation.md:32]: I moved it to the top level title "Customizing the R installation" so it's clear it's for all methods.

**Comment by frankier** [comment]
Updated with your comments. I guess the tests are false negatives? They don't seem to be to do with any of the changes and weren't there before (apart from Windows latest which was on master)?

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:56]: It would be good to note explicitly again that the preferences system can be used post-hoc and doesn't require a re-build.

Comment (palday) [reply docs/src/installation.md:56]: Actually, you have to call `Pkg.build()` again to activate the preferences file if you've previously built using a global install, right?

File: src/RCall.jl
Comment (palday) [inline src/RCall.jl:33]: ah this `@static` block means you need to do another `precompile` if you go from global R to preferences-R. But if the RCall source hasn't changed, then this won't work. Likewise, changing the preferences-R path won't be reflected. It's probably better to handle some of this logic as part of `__init__`: it will increase load-time slightly (but then again, you still have to start R afterwards, so a few milliseconds finding a path dynamically is nothing).

**Comment by palday** [comment]
@frankier I'm starting to understand more of [your concerns](https://github.com/JuliaInterop/RCall.jl/pull/496#issuecomment-1677540429) a bit more. It looks like there are a few hairy details to work out still -- I think it's going to be hard to mix performance/precompilation/dynamism. I'll try to work out some more concrete suggestions than "you might have to use `__init__`", but that takes a little bit more time. 🙁 

All that said, I think this is a valuable contribution. I'm not trying to nitpick but rather help you make your work shine. We just have a little bit more work to do.

File: src/RCall.jl
Comment (frankier) [reply src/RCall.jl:33]: I don't think that's needed. Have you tested this? I have a test script at https://gist.github.com/frankier/0ac468305f8b04d98c90551eda0da15d you can try. Just set RCALL_PATH to your checkout of this PR and run in from an empty directory. I see the following output:

```
...
Should be Conda R
/home/frankier/.julia/conda/3/lib/R/lib/libR.so
...
Should be CondaPkg R
/path/to/mytest/testproj/.CondaPkg/env/lib/R/lib/libR.so
...
```

The reason this works is whenever we read from a preference at compile time, it is marked as a compile time preference. I think this means that all bytecodes are then keyed off the preference value. The "global install" is keyed as `Rhome = libR = unset`. Then as soon as we set them we get a new key for our bytecode which triggers precompilation. I may have got the details wrong, but that's what I have surmised from the documentation.

**Comment by frankier** [comment]
I have addressed the doc comments. For the overall approach, see my reply above.

Just to note I did take a look at trying to load stuff later on in `__init__`. This seemed like it would need quite a lot of restructuring however, since stuff from libR is used at the top level in this package so this would need quite a big restructuring. It may be marginally useful still to have a module not hardwired to a particular libR e.g. for multiple R interpreters of different versions within a particular project which could be useful if you had different independent wrapper libraries using different R code, but as far as I have seen so far it looks like this preferences approach covers the common case fairly nicely.

**Comment by frankier** [comment]
@palday Have I addressed your concerns? Let me know if there's anything I can do to help with getting this merged.

**Comment by palday** [comment]
@frankier Can you add another few hours to the day? 😉 😭 Mostly I just need some time to really think about this and test it out and make sure that the behavior is something we want to commit to. I like the idea, but haven't gone through all the details yet. My apologies for the delay -- it's a really nice contribution, but non trivial, which means it needs a non trivial amount of thought to review!

**Comment by frankier** [comment]
Okay! Very understandable. Just checking it hadn't got lost in the mix. Cheers!

**Comment by frankier** [comment]
Okay -- I've added tests for installation including switching from environment variables to preferences as an automated test. It's a bit grotty but as far as I can see this is the best way to make sure it's run as part of the test suite. Hopefully this gives some extra assurances about the approach.

**Comment by schlichtanders** [comment]
@simonbyrne I am too looking for CondaPkg support for RCall - this pull request is the first needed for this. Can you please take a look and review the pullrequest?

**Comment by schlichtanders** [comment]
@frankier @palday can this be merged soon?

**Comment by frankier** [comment]
I can't merge this myself. I'm open to any feedback on improving the PR including adding more QA. It does have automated tests now.

**Comment by schlichtanders** [comment]
A question: Does this pullrequest enable setting RHome dynamically? i.e. that R does not need to be rebuild of RHome is changed?

**Comment by frankier** [comment]
This PR is designed to be as non-invasive as possible. Currently for various reasons including (IIRC) the implementation of the REPL mode, RCall.jl currently needs libR.so to compile. For this reason the Preferences system is used to create a compile-time preference. Compile time preferences give us the exact behavior we want: it will automatically retrigger a package compilation when the libR location is changed and it will keep different compiled copies for different libRs.

As it is, I would say this PR improves on the current situation, but libR still cannot be changed at runtime (I guess this is what is meant by dynamic), which might be preferable if possible. The alternative of making libR not needed at compile time might need a fairly big refactor possibly resulting in multiple packages. I don't have the resources for this myself, and I don't think the development resources seem readily forthcoming from elsewhere.

**Comment by frankier** [comment]
@palday I don't want to nag, nor to overstep, but I was wondering whether it might be possible to expedite this by finding someone else with commit access to this repo willing to review + merge this or to merge it yourself following a final check from some other trusted party? (In the latter case, I can try to find someone myself.)

**Comment by palday** [comment]
> @palday I don't want to nag, nor to overstep, 

Definitely not nagging and frankly your patience with my very limited attention is something I wish all contributors would have. :heart: Truly the right mix of patience and 'bump'ing!

My sincere apologies for not being a better shepherd here -- I haven't had the mental space for big FOSS work lately, especially for things that aren't part of my day job. :slightly_frowning_face: 

> but I was wondering whether it might be possible to expedite this 1) by finding someone else with commit access to this repo willing to review + merge this OR 2) to merge it yourself following a final check from some other trusted party? (In the latter case, I can try to find someone myself.)

Having a second opinion would definitely make me feel more comfortable. I can ping @simonbyrne, @randy3k and @dmbates as erst while maintainers of this package, but I don't know how much time/attention they have to spare at the moment.

File: src/setup.jl
Comment (palday) [inline src/setup.jl:176]:
```suggestion
            "No R installation was detected at RCall installation time. " *
```

**Comment by frankier** [comment]
Hi @palday! Thanks for asking in case anyone was around. I also asked on Slack and did not get a response. It looks like we're own our own!

I wanted to post a link to this talk, in particular "principle 9" given here: https://youtu.be/fQJrEQIRPZU?t=3547

What do you think? I have "skin in the game" in that I am using R packages quite a bit together with Julia, and am likely to continue to for quite a while since it's part of the ecosystem I have no desire to reimplement from scratch.

While I think there are no obvious problems, I understand your reservations since it's really easy to break people's workflows especially with the installation stage. What I can agree to do is to monitor this issues and try to address them, adding more tests as I go. To this end, I can also add an issue template mentioning that I can be pinged for any issues to do with installation or configuring which R installation to use.

To summarise one more time the reasons I think this PR should be merged

 1) **Makes sure that telling RCall.jl to use different R installations for different projects doesn't mean that projects accidentally clobber each other's preferences** (I think this is quite important, since it can cause pretty confusing problems for people)
 2) Allows the location of R to be specified after installation
 3) Allows having libR.so in a non-standard location see https://github.com/JuliaInterop/RCall.jl/issues/492
 4) Following on from 1 and 2, allows some kind of integration with RCall.jl, although as-per https://github.com/JuliaPy/CondaPkg.jl/issues/121 this may continue to require a manual step in some cases at least for the time being.

Even given the proviso of 4, I believe the changes made in this PR are still something we would want to integrate when making further improvements, since preferences should usually be the preferred way of configuring packages and overriding their defaults. I will probably continue to look into improving things after this PR is merged.

File: test/installation.jl
Comment (palday) [inline test/installation.jl:1]: I like testing the installation in more depth :smile:

File: test/installation.jl
Comment (palday) [inline test/installation.jl:125]:
```suggestion
# https://github.com/JuliaLang/julia/issues/34276
# (related to incompatible libstdc++ versions)
```

I used to run into this issue all the time....

File: test/runtests.jl
Comment (palday) [inline test/runtests.jl:4]:
```suggestion
@testset "installation" begin 
    include("installation.jl")
end
```

File: test/installation.jl
Comment (frankier) [reply test/installation.jl:1]: Great! Hopefully it addresses some of the possible failure modes you were bringing up before as well as making sure we can make future improvements. It should be reasonable easy to add more now the test harness is there in case you have any more ideas.

Comment (palday) [reply test/installation.jl:1]: I have a way to simplify this test set a fair amount -- do you mind if I push directly to your PR? It's a bit much to turn into a bunch of code suggestions.

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:14]: nit: this is at _build_ time for RCall. You can rebuild a package that's already been installed. (IIRC TimeZones also takes advantage of this to make its data are up to date.)

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:30]: It might be good to add a note here with the method used in the installation tests as a way to use CondaPkg-packaged R as the preferred route.

File: src/RCall.jl
Comment (palday) [inline src/RCall.jl:35]: not in this PR, but I would like to see the ability to figure out well-behaved libR when only Rhome is set

File: src/RCall.jl
Comment (palday) [inline src/RCall.jl:49]: it's very rare to see explicit `__precompile__` directives nowadays -- what happens if you don't have this?

File: src/RCall.jl
Comment (palday) [reply src/RCall.jl:33]: :thinking: It still feels weird to have a `@static` call for something that's dynamic. My preference would be drop it here because 
- it feels weird to me
- `@static` at the top-level of a module for something that's not platform dependent, doesn't actually get you that much (because the module definition is only evaluated once unless you do something fancy like explicitly invoking `Base.require` or wonky with Revise.jl)

But that's just my preference.

Comment (frankier) [reply src/RCall.jl]: That's a good point! I'm not sure why I set that, except I wanted to make extra-sure things would be at compile time, but as you say there's no need and as a result it's a bit odd. I have added a comment about what is going on here and pulled them out into temporary variables to defeat short-circuiting.

File: docs/src/installation.md
Comment (frankier) [reply docs/src/installation.md:14]: I suppose I was trying to contrast the two from the user's perspective, so that I could use consistent shorthand on this page. In a sense talking about build-time instead would also be misleading, since in terms of being a correct description of the implementation, R_HOME and the Preferences approach are *both* configured at build-time, but Preferences has the advantage that it triggers a build at the correct time and keeps multiple builds around. I have switched to using the `R_HOME` environment variable as a shorthand.

File: docs/src/installation.md
Comment (frankier) [reply docs/src/installation.md:30]: Good idea. Something better should be done here at some point, but I think a "cookbook"-style note is probably best for the meantime.

File: src/RCall.jl
Comment (frankier) [reply src/RCall.jl:40]: I guess having a default would be fine, but yes please, let's defer it for now.

File: src/RCall.jl
Comment (frankier) [reply src/RCall.jl:54]: This is to abort precompilation when there is not R available so that it is possible to install RCall.jl without an error when there is no R available, in case the user wants to set up preferences between installation and RCall import time. The reason for aborting precompilation is actually more about the dependent packages, since when they are precompiled they will import RCall and `__init__` will run and not be able to complete. In case this behaviour is surprising, see: https://github.com/frankier/precompiletest

I've added a message to this effect and re-written the code to make it more clear that this is about aborting precompilation rather than the old "opting-in" to precompilation approach.

**Comment by frankier** [comment]
Thanks for cleaning up the tests and for the feedback.

It looks like this test is failing https://github.com/JuliaInterop/RCall.jl/actions/runs/7444228626/job/20250301671?pr=496

Last time I looked it was failing due to a problem present on `main`, but now it seems to be an actual error. I guess the location of `R.dll` is being guessed incorrectly? Do you have any ideas? I guess it wasn't introduced during the test refactoring? I don't have a convenient way to test with Windows locally, but will try to set-up/fire-up a VM or something to troubleshoot in case the answer to those two questions is a negative.

**Comment by palday** [comment]
> Thanks for cleaning up the tests and for the feedback.
> 
> It looks like this test is failing https://github.com/JuliaInterop/RCall.jl/actions/runs/7444228626/job/20250301671?pr=496
> 
> Last time I looked it was failing due to a problem present on `main`, but now it seems to be an actual error. I guess the location of `R.dll` is being guessed incorrectly? Do you have any ideas? I guess it wasn't introduced during the test refactoring? I don't have a convenient way to test with Windows locally, but will try to set-up/fire-up a VM or something to troubleshoot in case the answer to those two questions is a negative.

Same here -- I've got a Windows VM set up, but I'm really slow figuring it out. AFAICT the path is correct, but there's something about loading DLLs not in the system variable `%PATH%` that doesn't work. I haven't yet figured it out, but I'm working on it when I get a few minutes in the evenings. :slightly_frowning_face: 

There was also an issue with the CI on macOS, but it looked to be a more fundamental issue with Conda and the availability of `md5` vs `md5sum` vs something else, so I just disabled that check. The last bit to get this merged is understanding the limitations on different platforms (e.g. "Linux can handle all this dynamically, but Windows requires setting XXX") and having the CI reflect that. Once we have that, I think this is good to go! So close!

**Comment by frankier** [comment]
Hi! I just checked this on a Windows laptop I got access to. Using dependencywalker.exe, I found that BLAS wasn't being found which led me to realise that this was actually a bug in the test code that was only surfacing on Windows. The CondaPkg environment is not activated(!) Apparently on the other systems another version of BLAS just gets picked up. On non-Conda R it's not a problem because everything is compiled into the DLLs I think. I fixed this and added a note to the docs. I also marked the information about using with CondaPkg as experimental.

**Comment by frankier** [comment]
Okay -- another failure on Windows. I feel like it worked correctly locally but I will look into it soon.

**Comment by palday** [comment]
> Okay -- another failure on Windows. I feel like it worked correctly locally but I will look into it soon.

I feel like we're heading in the right direction -- thanks for your patience and efforts!

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:58]: you can still do `using`, but you'll need to do

```julia
CondaPkg.withenv() do
    @eval using RCall
end
```

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:86]:
```suggestion
```

File: docs/src/installation.md
Comment (palday) [inline docs/src/installation.md:83]: this would actually work for any string that is not a path nor a `*`, right?

File: test/installation/swap_to_prefs_and_condapkg.jl
Comment (frankier) [inline test/installation/swap_to_prefs_and_condapkg.jl:42]: Is this still needed?

File: src/setup.jl
Comment (frankier) [inline src/setup.jl:175]: Ha! Mixed feelings about this sneaky hack to pass coverage. The reason there is no coverage here is because when we shell out in `installation.jl` coverage stops being measured. It's probably not worth tracking it.

**Comment by frankier** [comment]
Thanks for the docs feedback. I just left a couple of small double-checking comments back and then I guess merge time?

File: test/installation/swap_to_prefs_and_condapkg.jl
Comment (palday) [reply test/installation/swap_to_prefs_and_condapkg.jl:42]: It seemed to be in my testing.

File: src/setup.jl
Comment (palday) [reply src/setup.jl:175]: Yeah, it's definitely a hack but it's one I'm comfortable with for such "defense in depth" type measures.

**Review by palday — APPROVED** [review]

**Pull request merged**
